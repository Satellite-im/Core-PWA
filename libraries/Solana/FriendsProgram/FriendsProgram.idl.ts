export type Friends = {
  version: '0.1.0'
  name: 'friends'
  instructions: [
    {
      name: 'makeRequest'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
        {
          name: 'payer'
          isMut: true
          isSigner: true
        },
        {
          name: 'systemProgram'
          isMut: false
          isSigner: false
        },
      ]
      args: [
        {
          name: 'user1'
          type: 'publicKey'
        },
        {
          name: 'user2'
          type: 'publicKey'
        },
        {
          name: 'k'
          type: 'string'
        },
      ]
    },
    {
      name: 'acceptRequest'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
      ]
      args: [
        {
          name: 'k'
          type: 'string'
        },
      ]
    },
    {
      name: 'denyRequest'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
      ]
      args: []
    },
    {
      name: 'removeRequest'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
      ]
      args: []
    },
    {
      name: 'closeRequest'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
        {
          name: 'payer'
          isMut: true
          isSigner: false
        },
      ]
      args: []
    },
    {
      name: 'removeFriend'
      accounts: [
        {
          name: 'request'
          isMut: true
          isSigner: false
        },
        {
          name: 'user'
          isMut: false
          isSigner: true
        },
      ]
      args: []
    },
  ]
  accounts: [
    {
      name: 'friendRequest'
      type: {
        kind: 'struct'
        fields: [
          {
            name: 'from'
            type: 'publicKey'
          },
          {
            name: 'status'
            type: {
              defined: 'Status'
            }
          },
          {
            name: 'to'
            type: 'publicKey'
          },
          {
            name: 'payer'
            type: 'publicKey'
          },
          {
            name: 'fromEncryptedKey'
            type: 'string'
          },
          {
            name: 'toEncryptedKey'
            type: 'string'
          },
        ]
      }
    },
  ]
  types: [
    {
      name: 'Status'
      type: {
        kind: 'enum'
        variants: [
          {
            name: 'Uninitilized'
          },
          {
            name: 'Pending'
          },
          {
            name: 'Accepted'
          },
          {
            name: 'Denied'
          },
          {
            name: 'RemovedFriend'
          },
          {
            name: 'RequestRemoved'
          },
        ]
      }
    },
  ]
  errors: [
    {
      code: 6000
      name: 'WrongRequestData'
      msg: "Addresses in request don't match user address"
    },
    {
      code: 6001
      name: 'NotPendingRequest'
      msg: 'Request is not pending'
    },
    {
      code: 6002
      name: 'NotFriends'
      msg: 'Accounts are not friends yet'
    },
    {
      code: 6003
      name: 'WrongPrivileges'
      msg: "User can't perform this action"
    },
    {
      code: 6004
      name: 'OrderMismatch'
      msg: 'User1 and user2 needs to be passed in order'
    },
    {
      code: 6005
      name: 'AlreadyFriends'
      msg: 'Users are already friends'
    },
    {
      code: 6006
      name: 'ExistentRequest'
      msg: 'Request already existent'
    },
    {
      code: 6007
      name: 'PayerMismatch'
      msg: 'Account was not created by provided user'
    },
    {
      code: 6008
      name: 'NotRemoved'
      msg: 'Request is not removed yet'
    },
    {
      code: 6009
      name: 'AlreadyRemoved'
      msg: 'Request is already removed'
    },
  ]
}

export const IDL: Friends = {
  version: '0.1.0',
  name: 'friends',
  instructions: [
    {
      name: 'makeRequest',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
        {
          name: 'payer',
          isMut: true,
          isSigner: true,
        },
        {
          name: 'systemProgram',
          isMut: false,
          isSigner: false,
        },
      ],
      args: [
        {
          name: 'user1',
          type: 'publicKey',
        },
        {
          name: 'user2',
          type: 'publicKey',
        },
        {
          name: 'k',
          type: 'string',
        },
      ],
    },
    {
      name: 'acceptRequest',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
      ],
      args: [
        {
          name: 'k',
          type: 'string',
        },
      ],
    },
    {
      name: 'denyRequest',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
      ],
      args: [],
    },
    {
      name: 'removeRequest',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
      ],
      args: [],
    },
    {
      name: 'closeRequest',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
        {
          name: 'payer',
          isMut: true,
          isSigner: false,
        },
      ],
      args: [],
    },
    {
      name: 'removeFriend',
      accounts: [
        {
          name: 'request',
          isMut: true,
          isSigner: false,
        },
        {
          name: 'user',
          isMut: false,
          isSigner: true,
        },
      ],
      args: [],
    },
  ],
  accounts: [
    {
      name: 'friendRequest',
      type: {
        kind: 'struct',
        fields: [
          {
            name: 'from',
            type: 'publicKey',
          },
          {
            name: 'status',
            type: {
              defined: 'Status',
            },
          },
          {
            name: 'to',
            type: 'publicKey',
          },
          {
            name: 'payer',
            type: 'publicKey',
          },
          {
            name: 'fromEncryptedKey',
            type: 'string',
          },
          {
            name: 'toEncryptedKey',
            type: 'string',
          },
        ],
      },
    },
  ],
  types: [
    {
      name: 'Status',
      type: {
        kind: 'enum',
        variants: [
          {
            name: 'Uninitilized',
          },
          {
            name: 'Pending',
          },
          {
            name: 'Accepted',
          },
          {
            name: 'Denied',
          },
          {
            name: 'RemovedFriend',
          },
          {
            name: 'RequestRemoved',
          },
        ],
      },
    },
  ],
  errors: [
    {
      code: 6000,
      name: 'WrongRequestData',
      msg: "Addresses in request don't match user address",
    },
    {
      code: 6001,
      name: 'NotPendingRequest',
      msg: 'Request is not pending',
    },
    {
      code: 6002,
      name: 'NotFriends',
      msg: 'Accounts are not friends yet',
    },
    {
      code: 6003,
      name: 'WrongPrivileges',
      msg: "User can't perform this action",
    },
    {
      code: 6004,
      name: 'OrderMismatch',
      msg: 'User1 and user2 needs to be passed in order',
    },
    {
      code: 6005,
      name: 'AlreadyFriends',
      msg: 'Users are already friends',
    },
    {
      code: 6006,
      name: 'ExistentRequest',
      msg: 'Request already existent',
    },
    {
      code: 6007,
      name: 'PayerMismatch',
      msg: 'Account was not created by provided user',
    },
    {
      code: 6008,
      name: 'NotRemoved',
      msg: 'Request is not removed yet',
    },
    {
      code: 6009,
      name: 'AlreadyRemoved',
      msg: 'Request is already removed',
    },
  ],
}
