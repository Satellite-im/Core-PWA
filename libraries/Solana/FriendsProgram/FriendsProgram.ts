import {
  SystemProgram,
  Transaction,
  TransactionInstruction,
  PublicKey,
  sendAndConfirmTransaction,
  SYSVAR_RENT_PUBKEY,
  Keypair,
  // eslint-disable-next-line import/named
  ConfirmOptions,
} from '@solana/web3.js'
import base58 from 'micro-base58'
import { encodeInstructionData, friendLayout } from './FriendsProgram.layout'
import { CreateFriendParams, FriendStatus } from './FriendsProgram.types'

import { Seeds, publicKeyFromSeeds } from '~/libraries/Solana/Solana'

import Solana from '~/libraries/Solana/SolanaManager/SolanaManager'
import { Config } from '~/config'

export const FRIENDS_PROGRAM_ID = new PublicKey(Config.solana.friendsProgramId)

export default class FriendsProgram {
  solana: Solana
  constructor(solana: Solana) {
    this.solana = solana
  }

  async createDerivedAccount(
    seedKey: PublicKey,
    seedString: string,
    params: CreateFriendParams,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const { base, key } = await publicKeyFromSeeds(
      [seedKey.toBytes(), params.createAccount.friend.friendKey],
      seedString,
      FRIENDS_PROGRAM_ID
    )

    const instruction = new TransactionInstruction({
      keys: [
        { pubkey: payerAccount.publicKey, isSigner: true, isWritable: true },
        { pubkey: seedKey, isSigner: false, isWritable: false },
        { pubkey: base[0], isSigner: false, isWritable: false },
        { pubkey: key, isSigner: false, isWritable: true },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
        { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData(params),
    })

    const transaction = new Transaction().add(instruction)

    await sendAndConfirmTransaction(connection, transaction, [payerAccount], {
      commitment: 'finalized',
      preflightCommitment: 'finalized',
      ...confirmOptionsOverride,
    })
    return key
  }

  async createFriend(
    userFromKey: PublicKey,
    userToKey: PublicKey,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null
    if (!userFromKey) return null

    const params = {
      createAccount: {
        friend: { friendKey: userToKey.toBytes() },
      },
    }
    const friendKey = await this.createDerivedAccount(
      userFromKey,
      Seeds.Friend,
      params,
      confirmOptionsOverride
    )

    if (!friendKey) throw new Error('Derived account error')

    return friendKey
  }

  initFriendRequest(
    friendKey: PublicKey,
    friend2Key: PublicKey,
    userFromKey: PublicKey,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array
  ) {
    return new TransactionInstruction({
      keys: [
        { pubkey: friendKey, isSigner: false, isWritable: true },
        { pubkey: friend2Key, isSigner: false, isWritable: true },
        { pubkey: userFromKey, isSigner: true, isWritable: false },
        { pubkey: userToKey, isSigner: false, isWritable: true },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData({
        makeRequest: {
          tex: [
            fromPaddedBuffer.slice(0, 32),
            fromPaddedBuffer.slice(32, 64),
            toPaddedBuffer.slice(0, 32),
            toPaddedBuffer.slice(32, 64),
          ],
        },
      }),
    })
  }

  initAcceptFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array
  ) {
    return new TransactionInstruction({
      keys: [
        { pubkey: friendKey, isSigner: false, isWritable: true },
        { pubkey: userFromKey, isSigner: false, isWritable: true },
        { pubkey: userToKey, isSigner: true, isWritable: false },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData({
        acceptRequest: {
          tex: [
            fromPaddedBuffer.slice(0, 32),
            fromPaddedBuffer.slice(32, 64),
            toPaddedBuffer.slice(0, 32),
            toPaddedBuffer.slice(32, 64),
          ],
        },
      }),
    })
  }

  initDenyFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array
  ) {
    return new TransactionInstruction({
      keys: [
        { pubkey: friendKey, isSigner: false, isWritable: true },
        { pubkey: userFromKey, isSigner: false, isWritable: true },
        { pubkey: userToKey, isSigner: true, isWritable: false },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData({
        denyRequest: {
          tex: [
            fromPaddedBuffer.slice(0, 32),
            fromPaddedBuffer.slice(32, 64),
            toPaddedBuffer.slice(0, 32),
            toPaddedBuffer.slice(32, 64),
          ],
        },
      }),
    })
  }

  initRemoveFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array
  ) {
    return new TransactionInstruction({
      keys: [
        { pubkey: friendKey, isSigner: false, isWritable: true },
        { pubkey: userFromKey, isSigner: true, isWritable: false },
        { pubkey: userToKey, isSigner: false, isWritable: true },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData({
        removeRequest: {
          tex: [
            fromPaddedBuffer.slice(0, 32),
            fromPaddedBuffer.slice(32, 64),
            toPaddedBuffer.slice(0, 32),
            toPaddedBuffer.slice(32, 64),
          ],
        },
      }),
    })
  }

  initRemoveFriend(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array
  ) {
    return new TransactionInstruction({
      keys: [
        { pubkey: friendKey, isSigner: false, isWritable: true },
        { pubkey: userFromKey, isSigner: true, isWritable: false },
        { pubkey: userToKey, isSigner: false, isWritable: true },
        { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
      ],
      programId: FRIENDS_PROGRAM_ID,
      data: encodeInstructionData({
        removeFriend: {
          tex: [
            fromPaddedBuffer.slice(0, 32),
            fromPaddedBuffer.slice(32, 64),
            toPaddedBuffer.slice(0, 32),
            toPaddedBuffer.slice(32, 64),
          ],
        },
      }),
    })
  }

  createFriendRequest(
    friendKey: PublicKey,
    friend2Key: PublicKey,
    userFromAccount: Keypair,
    userToKey: PublicKey,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const transaction = new Transaction().add(
      this.initFriendRequest(
        friendKey,
        friend2Key,
        userFromAccount.publicKey,
        userToKey,
        fromPaddedBuffer,
        toPaddedBuffer
      )
    )

    return sendAndConfirmTransaction(
      connection,
      transaction,
      [payerAccount, userFromAccount],
      {
        commitment: 'singleGossip',
        preflightCommitment: 'singleGossip',
        ...confirmOptionsOverride,
      }
    )
  }

  acceptFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToAccount: Keypair,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const transaction = new Transaction().add(
      this.initAcceptFriendRequest(
        friendKey,
        userFromKey,
        userToAccount.publicKey,
        fromPaddedBuffer,
        toPaddedBuffer
      )
    )

    return sendAndConfirmTransaction(
      connection,
      transaction,
      [payerAccount, userToAccount],
      {
        commitment: 'singleGossip',
        preflightCommitment: 'singleGossip',
        ...confirmOptionsOverride,
      }
    )
  }

  denyFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToAccount: Keypair,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const transaction = new Transaction().add(
      this.initDenyFriendRequest(
        friendKey,
        userFromKey,
        userToAccount.publicKey,
        fromPaddedBuffer,
        toPaddedBuffer
      )
    )

    return sendAndConfirmTransaction(
      connection,
      transaction,
      [payerAccount, userToAccount],
      {
        commitment: 'singleGossip',
        preflightCommitment: 'singleGossip',
        ...confirmOptionsOverride,
      }
    )
  }

  removeFriendRequest(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToAccount: Keypair,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const transaction = new Transaction().add(
      this.initRemoveFriendRequest(
        friendKey,
        userFromKey,
        userToAccount.publicKey,
        fromPaddedBuffer,
        toPaddedBuffer
      )
    )

    return sendAndConfirmTransaction(
      connection,
      transaction,
      [payerAccount, userToAccount],
      {
        commitment: 'singleGossip',
        preflightCommitment: 'singleGossip',
        ...confirmOptionsOverride,
      }
    )
  }

  removeFriend(
    friendKey: PublicKey,
    userFromKey: PublicKey,
    userToAccount: Keypair,
    fromPaddedBuffer: Uint8Array,
    toPaddedBuffer: Uint8Array,
    confirmOptionsOverride?: ConfirmOptions
  ) {
    const { connection } = this.solana
    const payerAccount = this.solana.getActiveAccount()
    if (!payerAccount) return null

    const transaction = new Transaction().add(
      this.initRemoveFriend(
        friendKey,
        userFromKey,
        userToAccount.publicKey,
        fromPaddedBuffer,
        toPaddedBuffer
      )
    )

    return sendAndConfirmTransaction(
      connection,
      transaction,
      [payerAccount, userToAccount],
      {
        commitment: 'singleGossip',
        preflightCommitment: 'singleGossip',
        ...confirmOptionsOverride,
      }
    )
  }

  async getFriend(friendKey: PublicKey) {
    const { connection } = this.solana
    const accountInfo = await connection.getAccountInfo(friendKey)

    if (accountInfo === null) {
      return null
      //   throw new Error('Error: cannot find the account')
    }
    return friendLayout.decode(Buffer.from(accountInfo.data))
  }

  async computeFriendAccountKey(userFromKey: PublicKey, userToKey: PublicKey) {
    const { key } = await publicKeyFromSeeds(
      [userFromKey.toBytes(), userToKey.toBytes()],
      Seeds.Friend,
      FRIENDS_PROGRAM_ID
    )

    return key
  }

  async getFriendRequests(status: FriendStatus) {
    const { connection } = this.solana

    const userAccount = this.solana.getUserAccount()
    if (!userAccount) {
      throw new Error('User account not found')
    }

    const fromKeyAndStatus = base58(
      Buffer.from([...userAccount.publicKey.toBytes(), status])
    )

    const statusAndToKey = base58(
      Buffer.from([status, ...userAccount.publicKey.toBytes()])
    )

    const resultFrom = await connection.getParsedProgramAccounts(
      FRIENDS_PROGRAM_ID,
      {
        filters: [{ memcmp: { offset: 0, bytes: fromKeyAndStatus } }],
      }
    )

    const resultTo = await connection.getParsedProgramAccounts(
      FRIENDS_PROGRAM_ID,
      {
        filters: [{ memcmp: { offset: 32, bytes: statusAndToKey } }],
      }
    )

    return [...resultFrom, ...resultTo]
  }
}
