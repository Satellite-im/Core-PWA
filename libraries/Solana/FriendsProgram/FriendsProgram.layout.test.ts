import * as FriendsProgramLayout from '~/libraries/Solana/FriendsProgram/FriendsProgram.layout'

describe('FriendsProgramLayout.encodeInstructionData', () => {
  test('0', () => {
    const inst: any = new Uint8Array([1, 1])
    const inst2: any = new Uint8Array([-100, 0])
    const inst3: any = new Uint8Array([-100, 1])
    const inst4: any = new Uint8Array([0, 0])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      acceptRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = FriendsProgramLayout.encodeInstructionData({
      removeRequest: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = FriendsProgramLayout.encodeInstructionData({
      removeFriend: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const inst: any = new Uint8Array([-1, 1])
    const inst2: any = new Uint8Array([0, -1])
    const inst3: any = new Uint8Array([-1, -100])
    const inst4: any = new Uint8Array([0, -100])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      makeRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const inst: any = new Uint8Array([0, -1])
    const inst2: any = new Uint8Array([100, -1])
    const inst3: any = new Uint8Array([0, -100])
    const inst4: any = new Uint8Array([0, -1])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      makeRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })
})

describe('FriendsProgramLayout.encodeInstructionData', () => {
  test('0', () => {
    const result: any = FriendsProgramLayout.encodeInstructionData({
      removeFriend: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const inst: any = new Uint8Array([0, 100])
    const inst2: any = new Uint8Array([100, 1])
    const inst3: any = new Uint8Array([1, 100])
    const inst4: any = new Uint8Array([1, -1])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      makeRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = FriendsProgramLayout.encodeInstructionData({
      denyRequest: {},
    })
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const inst: any = new Uint8Array([0, -100])
    const inst2: any = new Uint8Array([-100, -1])
    const inst3: any = new Uint8Array([-100, 1])
    const inst4: any = new Uint8Array([100, 1])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      makeRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const inst: any = new Uint8Array([-1, 100])
    const inst2: any = new Uint8Array([1, -1])
    const inst3: any = new Uint8Array([100, 0])
    const inst4: any = new Uint8Array([-1, -1])
    const object: any = [inst, inst2, inst3, inst4]
    const result: any = FriendsProgramLayout.encodeInstructionData({
      makeRequest: { tex: object },
    })
    expect(result).toMatchSnapshot()
  })
})
