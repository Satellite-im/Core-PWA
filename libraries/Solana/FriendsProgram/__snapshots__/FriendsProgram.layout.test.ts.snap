// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendsProgramLayout.encodeInstructionData 0 1`] = `
Object {
  "data": Array [
    1,
    1,
    1,
    156,
    0,
    156,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 0 2`] = `
Object {
  "data": Array [
    4,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 1 1`] = `
Object {
  "data": Array [
    3,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 1 2`] = `
Object {
  "data": Array [
    0,
    0,
    100,
    100,
    1,
    1,
    100,
    1,
    255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 2 1`] = `
Object {
  "data": Array [
    4,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 2 2`] = `
Object {
  "data": Array [
    2,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 3 1`] = `
Object {
  "data": Array [
    0,
    255,
    1,
    0,
    255,
    255,
    156,
    0,
    156,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 3 2`] = `
Object {
  "data": Array [
    0,
    0,
    156,
    156,
    255,
    156,
    1,
    100,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 4 1`] = `
Object {
  "data": Array [
    0,
    0,
    255,
    100,
    255,
    0,
    156,
    0,
    255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`FriendsProgramLayout.encodeInstructionData 4 2`] = `
Object {
  "data": Array [
    0,
    255,
    100,
    1,
    255,
    100,
    0,
    255,
    255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;
